### PR TITLE
Serve storage assets through CloudFront

### PR DESCRIPTION
**Pull request by MiroDojkic**
### This PR
- enables storage access using CloudFront
- fallbacks to own proxy endpoint: `GET /proxy/:key`
- access is verified with signed cookies
- cookies allow access to defined resources, i.e. `repository/access`
- cookies are obtained on repository endpoints: `ALL /repository/:id`

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
 'use strict';
 
 const bodyParser = require('body-parser');
+const cookieParser = require('cookie-parser');
 const cors = require('cors');
 const express = require('express');
 const helmet = require('helmet');
 const origin = require('./shared/origin');
 const path = require('path');
+const proxy = require('./shared/storage/proxy');
```
Comment (underscope) [inline server/app.js]: Please rename to `storageProxy` to be more exact (there could be other types of proxies in this type of file, includes readability in the following lines ;) )

File: client/api/asset.js [hunk 2]
```diff
@@ -1,9 +1,14 @@
 import request from './request';
 
 const urls = {
-  base: '/assets'
+  base: '/assets',
+  setCookies: '/assets/set-cookies'
```
Comment (underscope) [inline client/api/asset.js]: Does it make sense to be even more specific (i.e. set-access-cookies)

File: client/api/asset.js [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
 };
 
+function setCookies() {
```
Comment (underscope) [inline client/api/asset.js]: Same as prev; feels a bit too generic when called from outside (what kind of cookies 🤔 )

File: config/server/index.js [hunk 4]
```diff
@@ -15,6 +16,7 @@ const previewUrl = process.env.PREVIEW_URL;
 module.exports = {
   protocol,
   hostname,
+  localhost,
```
Comment (underscope) [inline config/server/index.js]: If flag, please keep the is prefix. Move on the first place(protocol, hostname, port looks like expected order/grouping)

File: config/server/storage.js [hunk 5]
```diff
@@ -10,5 +11,16 @@ module.exports = {
   filesystem: {
     path: process.env.STORAGE_PATH
   },
-  provider: process.env.STORAGE_PROVIDER
+  proxy: {
+    provider: process.env.STORAGE_PROXY,
```
Comment (underscope) [inline config/server/storage.js]: Default here to proxy for backwards comp?

File: server/shared/storage/helpers.js [hunk 6]
```diff
@@ -100,7 +101,7 @@ async function resolveAssetsMap(element) {
     if (!url) return set(element.data, key, url);
     const isStorageResource = url.startsWith(STORAGE_PROTOCOL);
     const resolvedUrl = isStorageResource
-      ? (await storage.getFileUrl(url.substr(STORAGE_PROTOCOL.length, url.length)))
+      ? proxy.getFileUrl(url.substr(STORAGE_PROTOCOL.length, url.length))
```
Comment (underscope) [inline server/shared/storage/helpers.js:104]: Omit async from f def.

File: server/shared/storage/helpers.js [hunk 7]
```diff
@@ -5,16 +5,17 @@ const { elementRegistry } = require('../content-plugins');
 const get = require('lodash/get');
 const isString = require('lodash/isString');
 const isUrl = require('is-url');
+const config = require('../../../config/server').storage;
 const mime = require('mime-types');
 const Promise = require('bluebird');
+const proxy = require('./proxy');
```
Comment (underscope) [inline server/shared/storage/helpers.js:11]: ✅ Here (within storage helpers)

File: server/shared/storage/storage.controller.js [hunk 8]
```diff
@@ -16,21 +15,10 @@ async function upload({ file }, res) {
   const hash = sha256(file.originalname, buffer);
   const extension = path.extname(file.originalname);
   const name = path.basename(file.originalname, extension).substring(0, 180).trim();
-  const key = path.join(ASSET_ROOT, `${hash}___${name}${extension}`);
+  const key = path.join(config.path, `${hash}___${name}${extension}`);
```
Comment (underscope) [inline server/shared/storage/storage.controller.js:18]: Lets consider to start scoping these per repo.

File: server/shared/storage/proxy/mw.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+const { FORBIDDEN } = require('http-status-codes');
+const miss = require('mississippi');
+const path = require('path');
+const proxy = require('./');
+const storage = require('../');
+const config = require('../../../../config/server').storage;
```
Comment (underscope) [inline server/shared/storage/proxy/mw.js]: Order

File: poi.config.js [hunk 10]
```diff
@@ -37,6 +38,7 @@ const devServer = {
   },
   proxy: {
     '/api': { target: serverUrl },
+    ...(STORAGE_PROXY === 'local' ? { '/proxy': serverUrl } : {}),
```
Comment (underscope) [inline poi.config.js]: Should the base URL for the proxy (local or cf) always be exposed in order to be able to construct asset url from client based on key?

**Comment by underscope** [comment]
🏆 🎖️

File: client/api/asset.js (hunk 2 above)
Comment (MiroDojkic) [reply client/api/asset.js]: Cookies are now set on the response for the `/repository/:id` request, so I removed this as it is no longer used.

File: config/server/index.js (hunk 4 above)
Comment (MiroDojkic) [reply config/server/index.js]: Removed this too since it's not used anymore.

**Review by underscope — APPROVED** [review]

**Pull request merged**